### PR TITLE
Up times graphs and availability 

### PR DESCRIPTION
**Pull request by Diogo-Ferreira**
Hi ! @Swisscom-bigdata we use Cachet as our status page for our platform. We need a missing feature on Cachet which is https://github.com/CachetHQ/Cachet/issues/1656 and so we decided to implement it and share it with you.

image

# Features
* Up and Downtime (Hourly and Daily) for each component and component groups
* Incident list when you click on the red bars
* Availability in the date range
* Date range
* CSV and Excel export

# How are the downtimes calculated?
An incident is only considered as downtime (red bar) when the component is in *major outage*. From the start time of the incident till the update with the *fixed* status. (Incidents without updates or without a fixed update will be red till the current date)

The calculations are made on the fly with SQL and PHP, there's only 1 migration which is adding a component_status to the incident table.

# Frontend code
For now, the frontend is in JQuery, but as Cachet moved to Vue.JS this is problematic. I found a quick solution by moving the script into a file and add it in webpack in the all.js file. This isn't an elegant solution of course, but if you're interested in this PR I will make time to convert it to Vue.JS.

# Possible improvements
* Configuration in the dashboard (Show or not the uptimes, colors etc...)
* Better code? (Honestly, I don't have a lot of experience with Laravel, maybe you guys can give me some hints how to improve the code)
* Other formats of exports (Plot.ly seems interesting).
* Other colors for Performance issues, partial outages...

### CODE REVIEW

**Comment by jbrooksuk** [comment]
Interesting, although this needs rebasing before we can properly review it, please :)

**Comment by ConnorVG** [comment]
That is a lot of file changes. Is this feature complete?

**Comment by Diogo-Ferreira** [comment]
@ConnorVG The configuration via the dashboard is not yet implemented, but the rest is complete.

File: .cfignore [hunk 1]
```diff
@@ -0,0 +1,2 @@
+/.git
```
Comment (jbrooksuk) [inline .cfignore]: Please don't add this file in.

File: .gitignore [hunk 2]
```diff
@@ -9,3 +9,6 @@ Homestead.json
 Homestead.yaml
 npm-debug.log
 .env
+.env.prod
```
Comment (jbrooksuk) [inline .gitignore]: Please don't add these.

File: VERSION [hunk 3]
```diff
@@ -1 +1 @@
-2.4.0-dev
+2.5.0-dev
```
Comment (jbrooksuk) [inline VERSION]: Please don't change the version, this is handled by the maintainers.

File: app/Console/Kernel.php [hunk 4]
```diff
@@ -39,6 +39,7 @@ class Kernel extends ConsoleKernel
         DemoSeederCommand::class,
         InstallCommand::class,
         VersionCommand::class,
+        \Laravel\Tinker\Console\TinkerCommand::class,
```
Comment (jbrooksuk) [inline app/Console/Kernel.php]: Please import and then use the class name here.

File: app/Foundation/Providers/ModuleServiceProvider.php [hunk 5]
```diff
@@ -44,9 +45,10 @@ class ModuleServiceProvider extends ServiceProvider
             'messages'   => 10000,
             'status'     => 20000,
             'components' => 30000,
+            'uptime'     => 35000,
             'metrics'    => 40000,
             'scheduled'  => 50000,
-            'stickied'   => 60000,
+            'stickied'   => 31000,
```
Comment (jbrooksuk) [inline app/Foundation/Providers/ModuleServiceProvider.php]: Why has this been pushed down?

File: app/Foundation/Providers/RepositoryServiceProvider.php [hunk 6]
```diff
@@ -58,4 +62,23 @@ protected function registerMetricRepository()
...
+     *
+     * @return void
+     */
+    protected function registerUpTimeRepository()
+    {
+        $this->app->singleton(UpTimeRepository::class, function (Container $app) {
+            $config = $app->make(ConfigRepository::class);
+
+            switch ($config->get('database.default')) {
+                case 'pgsql': $repository = new UpTimePgSqlRepository($config); break;
```
Comment (jbrooksuk) [inline app/Foundation/Providers/RepositoryServiceProvider.php]: What about SQLite?

File: app/Http/Kernel.php [hunk 7]
```diff
@@ -42,4 +42,12 @@ class Kernel extends HttpKernel
         'subscribers' => 'CachetHQ\Cachet\Http\Middleware\SubscribersConfigured',
         'throttle'    => 'AltThree\Throttle\ThrottlingMiddleware',
     ];
+
+    protected $middlewareGroups = [
```
Comment (jbrooksuk) [inline app/Http/Kernel.php]: This should be added to `RouteServiceProvider@mapApiRoutes`

File: app/Http/Kernel.php [hunk 8]
```diff
@@ -42,4 +42,12 @@ class Kernel extends HttpKernel
...
+        'api' => [
+            '\App\Http\Middleware\EncryptCookies::class',
```
Comment (jbrooksuk) [inline app/Http/Kernel.php]: We don't use this namespace.

File: app/Models/Component.php [hunk 9]
```diff
@@ -20,6 +20,9 @@
 use Illuminate\Database\Eloquent\SoftDeletes;
 use McCool\LaravelAutoPresenter\HasPresenter;
 
+/**
+ * @property mixed id
```
Comment (jbrooksuk) [inline app/Models/Component.php]: We don't add `@property` elsewhere, so I feel this is out of place.

File: app/Repositories/Uptime/AbstractUpTimeRepository.php [hunk 10]
```diff
@@ -0,0 +1,115 @@
...
+
+namespace CachetHQ\Cachet\Repositories\Uptime;
+
+use CachetHQ\Cachet\Models\IncidentUpdate;
+use Carbon\Carbon;
+use Illuminate\Config\Repository;
+
+class AbstractUpTimeRepository
+{
+    protected $config;
```
Comment (jbrooksuk) [inline app/Repositories/Uptime/AbstractUpTimeRepository.php]: Missing docblocks.

File: app/Repositories/Uptime/AbstractUpTimeRepository.php [hunk 11]
```diff
@@ -0,0 +1,115 @@
...
+
+    const DOWN_TIME_STATUSES = [4];
```
Comment (jbrooksuk) [inline app/Repositories/Uptime/AbstractUpTimeRepository.php]: Missing docblocks.

File: app/Repositories/Uptime/AbstractUpTimeRepository.php [hunk 12]
```diff
@@ -0,0 +1,115 @@
...
+    const FIXED_UPDATE_STATUS_ID = 4;
+
+    /**
+     * AbstractUpTimeRepository constructor.
+     *
+     * @param Repository $config
```
Comment (jbrooksuk) [inline app/Repositories/Uptime/AbstractUpTimeRepository.php]: Use FQN.

File: app/Repositories/Uptime/AbstractUpTimeRepository.php [hunk 13]
```diff
@@ -0,0 +1,115 @@
...
+     */
+    public function __construct(Repository $config)
+    {
+        $this->config = $config;
+    }
+
+    protected function getIncidentsTable()
```
Comment (jbrooksuk) [inline app/Repositories/Uptime/AbstractUpTimeRepository.php]: Missing docblocks all through here.

File: app/Repositories/Uptime/AbstractUpTimeRepository.php [hunk 14]
```diff
@@ -0,0 +1,115 @@
...
+                'min_date'          => $e->min_time,
+                'max_date'          => $e->max_time,
+            ];
+        })->values();
+
+        return compact('downTimeHours', 'incidents');
+    }
+
+    /**
+     * @param $row
```
Comment (jbrooksuk) [inline app/Repositories/Uptime/AbstractUpTimeRepository.php]: No method description.

File: app/Repositories/Uptime/UpTimeMySqlRepository.php [hunk 15]
```diff
@@ -0,0 +1,32 @@
...
+     * @param Collection $component
+     * @param $toDateEpoch
+     * @param bool $fromDateEpoch
+     *
+     * @return mixed
+     */
+    public function getComponentsIncidentsAndUpdates(Collection $components)
+    {
+        return DB::select(
+            'SELECT component_id,incidents.name, incidents.id as id,  UNIX_TIMESTAMP(max_time) as max_time, UNIX_TIMESTAMP (incidents.occurred_at) as min_time FROM ( SELECT incident_id, MAX(incident_updates.updated_at) as max_time FROM incident_updates JOIN incidents ON incident_id=incidents.id WHERE incident_updates.status = '.self::FIXED_UPDATE_STATUS_ID.' GROUP BY incident_id,incidents.occurred_at ) AS updates RIGHT JOIN incidents ON updates.incident_id = incidents.id WHERE component_id IN ( '.$components->implode('id', ',').' ) AND incidents.component_status IN ('.implode(',', self::DOWN_TIME_STATUSES).' )'
```
Comment (jbrooksuk) [inline app/Repositories/Uptime/UpTimeMySqlRepository.php]: Please use prepared statements.

File: app/Repositories/Uptime/UpTimeRepository.php [hunk 16]
```diff
@@ -0,0 +1,115 @@
...
+/**
+ * UpTime repository where we calculate the uptimes based on incidents.
+ *
+ * @author Ferreira Venancio Diogo <Diogo.FerreiraVenancio@swisscom.com>
+ */
+class UpTimeRepository
+{
+    private $repository;
+
+    private $userTimeZone = 'Europe/Berlin'; // TODO: adapt this automatically
```
Comment (jbrooksuk) [inline app/Repositories/Uptime/UpTimeRepository.php]: This can be `app.timezone`.

File: composer.json [hunk 17]
```diff
@@ -38,6 +38,7 @@
         "backup-manager/laravel": "dev-master#df53f9c9d8c6be5d7a2638f45d54b8fb7bc51e2b",
         "barryvdh/laravel-cors": "^0.8",
         "doctrine/dbal": "2.5.13",
+        "dompdf/dompdf": "0.6.1",
```
Comment (jbrooksuk) [inline composer.json]: Why do we have DOMPdf and MPdf?

File: composer.json [hunk 18]
```diff
@@ -47,11 +48,15 @@
         "jenssegers/date": "^3.2",
         "laravel/framework": "^5.4",
         "laravel/tinker": "^1.0",
+        "maatwebsite/excel": "^2.1",
```
Comment (jbrooksuk) [inline composer.json]: Why do we have excel and phpexcel?

File: composer.json [hunk 19]
```diff
@@ -47,11 +48,15 @@
...
         "mccool/laravel-auto-presenter": "^5.0",
+        "mpdf/mpdf": "6.1.0",
         "nexmo/client": "@beta",
+        "phpoffice/phpexcel": "^1.8",
         "pragmarx/google2fa": "^0.7.1",
         "predis/predis": "^1.1",
         "roumen/feed": "^2.10",
+        "tecnick.com/tcpdf": "~6.0",
```
Comment (jbrooksuk) [inline composer.json]: TCPDF too?

File: composer.json [hunk 20]
```diff
@@ -90,18 +95,23 @@
     },
     "scripts": {
         "post-root-package-install": [
-            "php -r \"file_exists('.env') || copy('.env.example', '.env');\""
+            "php -r \"copy('.env.example', '.env');\"",
```
Comment (jbrooksuk) [inline composer.json]: Please don't make changes here.

File: composer.json [hunk 21]
```diff
@@ -90,18 +95,23 @@
...
+            "php artisan key:generate"
```
Comment (jbrooksuk) [inline composer.json]: Don't add this.

File: composer.json [hunk 22]
```diff
@@ -90,18 +95,23 @@
...
         ],
         "post-create-project-cmd": [
             "php artisan key:generate"
         ],
         "post-install-cmd": [
             "Illuminate\\Foundation\\ComposerScripts::postInstall",
-            "php artisan optimize"
+            "php artisan optimize --force",
+            "php artisan config:cache",
```
Comment (jbrooksuk) [inline composer.json]: Don't make changes to these scripts.

File: composer.json [hunk 23]
```diff
@@ -111,7 +121,7 @@
     },
     "extra": {
         "branch-alias": {
-            "dev-master": "2.4-dev"
+            "dev-master": "2.5-dev"
```
Comment (jbrooksuk) [inline composer.json]: Again, not for you to change.

File: config/app.php [hunk 24]
```diff
@@ -73,7 +73,7 @@
     |
     */
 
-    'timezone' => 'UTC',
+    'timezone' => env('TIME_ZONE', 'UTC'),
```
Comment (jbrooksuk) [inline config/app.php]: This should be within its own PR.

File: config/database.php [hunk 25]
```diff
@@ -9,6 +9,26 @@
  * file that was distributed with this source code.
  */
 
+$cfEnv = @$_ENV['VCAP_SERVICES'];
```
Comment (jbrooksuk) [inline config/database.php]: This is unrelated to this change set.

File: config/mail.php [hunk 26]
```diff
@@ -38,7 +38,7 @@
     |
     */
 
-    'host' => env('MAIL_HOST', 'smtp.mailgun.org'),
+    'host' => env('MAIL_HOST', 'appmail.swisscom.com'),
```
Comment (jbrooksuk) [inline config/mail.php]: Don't change default values to an unknown service.

File: database/migrations/2017_09_15_090237_add_component_status_to_incidents.php [hunk 27]
```diff
@@ -0,0 +1,41 @@
...
+{
+    /**
+     * Run the migrations.
+     *
+     * @return void
+     */
+    public function up()
+    {
+        Schema::table('incidents', function (Blueprint $table) {
+            $table->integer('component_status')->default(1);
```
Comment (jbrooksuk) [inline database/migrations/2017_09_15_090237_add_component_status_to_incidents.php]: This needs to be `unsigned`.

File: options.json [hunk 28]
```diff
@@ -0,0 +1,5 @@
+{
```
Comment (jbrooksuk) [inline options.json]: What is this file for?

File: package.json [hunk 29]
```diff
@@ -35,5 +35,8 @@
     "sweetalert2": "^6.8.0",
     "vue": "^2.4.4"
   },
-  "private": true
+  "private": true,
+  "dependencies": {
+    "vue-loader": "^13.3.0"
```
Comment (jbrooksuk) [inline package.json]: Wrong place.

File: phpunit.xml.dist [hunk 30]
```diff
@@ -33,8 +33,8 @@
         <env name="APP_DEBUG" value="true"/>
         <env name="APP_URL" value="http://localhost"/>
         <env name="APP_KEY" value="GCvcgDKMRIN498g52zfVEd9CxDs6PR7q"/>
-        <env name="DB_DRIVER" value="sqlite"/>
-        <env name="DB_DATABASE" value=":memory:"/>
+        <env name="DB_DRIVER" value="pgsql"/>
```
Comment (jbrooksuk) [inline phpunit.xml.dist]: Don't change testing values.

File: public/dist/js/manifest.js [hunk 31]
```diff
@@ -1 +1,150 @@
...
\ No newline at end of file
+/******/ (function(modules) { // webpackBootstrap
```
Comment (jbrooksuk) [inline public/dist/js/manifest.js]: Compiled assets should not be part of a PR, we'll do this.

**Comment by Diogo-Ferreira** [comment]
Hi @jbrooksuk , thanks I will apply the changes that you requested. I will split this PR into smaller chunks too as your requested, **what do you think about**:
* 1 PR for UpTimes Controller and Repository
* 1 PR for the template and frontend code
* 1 PR for the charts exports to Excel and CSV 

Thank you in advance, sorry for the mess this is one of my first OS contributions :)

**Comment by jbrooksuk** [comment]
Sorry, one PR is fine for all of the above, but you've made several unrelated changes too.

File: app/Foundation/Providers/RepositoryServiceProvider.php (hunk 6 above)
Comment (Diogo-Ferreira) [reply app/Foundation/Providers/RepositoryServiceProvider.php]: Sqlite doesn't support right joins apparently. I'm thinking that maybe a repository isn't the good solution. If this can be done with the eloquent ORM it would be better. The problem is a subquery.

**Comment by comnam90** [comment]
Ooo Super keen to see this one make it's into Cachet

**Comment by jamesgraham** [comment]
Uptime is such a basic metric, is there any ETA on this?

**Comment by Diogo-Ferreira** [comment]
@jamesgraham Sorry, I'm busy for the end of the year and probably won't be able to continue until next year :(

**Comment by jbrooksuk** [comment]
We'll leave this open for you @Diogo-Ferreira.

**Comment by meonkeys** [comment]
Woah, cool! I didn't even know this one was in progress. :crossed_fingers:

**Comment by mpawlak96** [comment]
hi, how can i generate this update graphs and display on site ?

**Comment by bbarclay** [comment]
Is this going to become a feature? This would be awesome?

**Comment by gazben** [comment]
@Diogo-Ferreira Do you have time for finishing this? Do you need help?

**Comment by Diogo-Ferreira** [comment]
Hi @gazben, no not really and I don't know if I'll continue on this one. Help is very welcome :) ! Also I think there has been a lot of changes since the creation of this PR, for instance jquery is used but I think Cachet uses vue now, maybe it's better to start over ?

**Comment by gazben** [comment]
@Diogo-Ferreira Yes there are a lot of changes. We will see how my current project goes and if we need advanced features I will take a look at this

**Comment by alexstojda** [comment]
@jbrooksuk whats missing for this PR to be merged in? I'm interested in picking it up

**Comment by alexcurtin** [comment]
@jbrooksuk is something like this or similar being planned for the 2.4/3.0 branch?

**Comment by bbarclay** [comment]
You can link your database to graphana and then embed your graphs back into your system with an iframe.

**Comment by ValentinAUCLERC** [comment]
That's the only missing thing that prevent me to use it in my company :(

**Comment by ricardovf** [comment]
Any possibility to include this?

**Comment by jbrooksuk** [comment]
Thank you for your input on Cachet 2.x. We are shifting our attention and resources to Cachet 3.x and will no longer be supporting the 2.x version. If your feedback or issue is relevant to the 3.x series, we encourage you to engage with the new branch.

For more information on the Cachet rebuild and our plans for 3.x, you can read the announcement [here](https://github.com/CachetHQ/Cachet/discussions/4342).

We appreciate your understanding and look forward to your contributions to the new version.
